### DISCUSSION TITLE
polymorphic identity on insert? and insert/update->returning?

### DISCUSSION THREAD

**Discussion by jensimik** [sqlalchemy/sqlalchemy#6588] [Usage Questions] [answered]
See example below - do i use the polymorphic feature in a wrong way?

1) when using insert(B) - i would expect the polymorphic_identity to be set too?
2) when using insert/update with returning and select.from_statement i get an TypeError exception when it is a polymorphic - any options i miss?

sqlalchemy 1.4.17 / postgresql

```

import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()
engine = sa.create_engine(POSTGRES, echo=True, future=True)
Session = sa.orm.sessionmaker(bind=engine)


class A(Base):
    __tablename__ = "a"

    id = sa.Column(sa.Integer, sa.Identity(start=1, increment=1), primary_key=True)
    obj_type = sa.Column(sa.String)

    __mapper_args__ = {"polymorphic_identity": "a", "polymorphic_on": obj_type}


class B(A):
    name = sa.Column(sa.String)

    __mapper_args__ = {"polymorphic_identity": "b"}


if __name__ == "__main__":
    Base.metadata.create_all(engine)
    session = Session()

    q = sa.insert(B).values({"name": "b"})
    session.execute(q)
    # generated SQL is:
    # INSERT INTO a (name) VALUES (%(name)s) RETURNING a.id
    # would expect obj_type = "b" to be set too?

    q = sa.future.select(B).from_statement(
        sa.update(B).values({"name": "c"}).where(B.name == "b").returning(B)
    )
    b = session.execute(q).scalar_one()
    # TypeError: 'sqlalchemy.cimmutabledict.immutabledict' object does not support item assignment
```

**Comment by zzzeek** [comment, marked as answer]
> See example below - do i use the polymorphic feature in a wrong way?
> 
>     1. when using insert(B) - i would expect the polymorphic_identity to be set too?

no, the insert() construct has no ORM-enabled features, and in general insert/update/delete used directly do not have any inheritance/polymorphic behaviors.     This is a little bit unfortunate with 1.4's usage model that tries to have select/insert/update/delete work in ORM contexts, however for ORM polymorphic persistence features you'd need to use session.add().


> 
>     2. when using insert/update with returning and select.from_statement i get an TypeError exception when it is a polymorphic - any options i miss?


that's #6591 and will be fixed in the next release.
